HSEARCH-3552 Field type compatibility checks ignore the analyzer

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3552

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/impl/LuceneScopedIndexFieldComponent.java:
```diff
@@ -8,19 +8,32 @@
 		return component;
 	}
 
+	public void setComponent(T component) {
```
The idea would be the one to use the instance of `LuceneScopedIndexFieldComponent' as a collector, instead of using local variables.
Good idea to make the setters package visible only.

fax4ever commented:
@yrodiere thanks for the review. I think I've addressed the changes here.
It wouldn't be an issue to update the documentation by myself. For me it's the same.

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -1243,6 +1253,59 @@ public void multiIndex_withIncompatibleIndexManager_dslConverterDisabled() {
+						.analyzer( OverrideAnalysisDefinitions.ANALYZER_WHITESPACE_LOWERCASE.name ) )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -1243,6 +1253,59 @@ public void multiIndex_withIncompatibleIndexManager_dslConverterDisabled() {
+						.skipAnalysis() )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/PhraseSearchPredicateIT.java:
```diff
@@ -660,25 +660,57 @@ public void multiIndex_withRawFieldCompatibleIndexManager() {
+						.analyzer( OverrideAnalysisDefinitions.ANALYZER_WHITESPACE_LOWERCASE.name ) )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/PhraseSearchPredicateIT.java:
```diff
@@ -660,25 +660,57 @@ public void multiIndex_withRawFieldCompatibleIndexManager() {
+						.skipAnalysis() )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/SimpleQueryStringSearchPredicateIT.java:
```diff
@@ -683,24 +687,57 @@ public void multiIndex_withRawFieldCompatibleIndexManager() {
+						.analyzer( OverrideAnalysisDefinitions.ANALYZER_WHITESPACE_LOWERCASE.name ) )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/SimpleQueryStringSearchPredicateIT.java:
```diff
@@ -683,24 +687,57 @@ public void multiIndex_withRawFieldCompatibleIndexManager() {
+						.skipAnalysis() )
+				.toQuery();
+
+		assertThat( query ).assertThat( query ).hasDocRefHitsAnyOrder( b -> {
```
`assertThat( query ).assertThat( query )` ?

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/impl/LuceneSearchScopeModel.java:
```diff
@@ -76,49 +76,51 @@ public EventContext getIndexesEventContext() {
+				);
+			}
+
+			LuceneFailingCompatibilityChecker failingCompatibilityChecker = new LuceneFailingCompatibilityChecker(
```
Missing generics in both the variable type and the constructor invocation.

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/impl/ElasticsearchSearchScopeModel.java:
```diff
@@ -78,49 +78,51 @@ public EventContext getIndexesEventContext() {
+						)
+				);
+			}
+			ElasticsearchFailingCompatibilityChecker failingCompatibilityChecker = new ElasticsearchFailingCompatibilityChecker<>(
```
Missing generics in the variable type.

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/impl/ElasticsearchSearchScopeModel.java:
```diff
@@ -78,49 +78,51 @@ public EventContext getIndexesEventContext() {
+					absoluteFieldPath, scopedIndexFieldComponent.getComponent(), component, EventContexts.fromIndexNames(
+					indexModelForSelectedSchemaNode.getHibernateSearchIndexName(),
+					indexModel.getHibernateSearchIndexName()
+			), componentRetrievalStrategy );
```
This is indented quite strangely. Even automatic formatting in IDEA (which usually isn't great) looks better.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/impl/LuceneSearchScopeModel.java:
```diff
@@ -76,49 +76,51 @@ public EventContext getIndexesEventContext() {
+					absoluteFieldPath, scopedIndexFieldComponent.getComponent(), component, EventContexts.fromIndexNames(
+					indexModelForSelectedSchemaNode.getIndexName(),
+					indexModel.getIndexName()
+			), componentRetrievalStrategy );
```
This is indented quite strangely. Even automatic formatting in IDEA (which usually isn't great) looks better.

fax4ever commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchTextFieldPredicateBuilderFactory.java:
```diff
@@ -21,18 +21,33 @@
 		extends ElasticsearchStandardFieldPredicateBuilderFactory<String> {
 
 	private final DataType type;
+	private final String analyzerOrNormalizer;
```
Exactly. In fact, we're trying here to do what we did for the Lucene backend. I think that if we decided to change that, we should change  the Lucene backend as well.

fax4ever replied:
So shall we change that for both backends?

yrodiere replied:
Yes that sounds better.

yrodiere replied:
Note that I don't mean there should be a separate normalizerCompatibilityChecker: just having the analyzer name and normalizer name separate in this class and checking the compatibility of both in `hasCompatibleAnalyzer` should be enough.

fax4ever replied:
I think I got it. But let's wait my change to be sure about that :).

fax4ever commented:
@yrodiere thanks for the review. I think I've addressed all the comments here. Actually I've changed my mind on validating explicit analyzers and normalizers in Lucene backend too. The reason is that analyzerOrNormalizer field is very relative to several class, so if we wanted to change that, I think we could easily go out of the scope of the current issue.

yrodiere commented:
> The reason is that analyzerOrNormalizer field is very relative to several classes, so if we wanted to change that, I think we could easily go out of the scope of the current issue.

Two classes are involved: `org.hibernate.search.backend.lucene.types.predicate.impl.LuceneTextFieldPredicateBuilderFactory` and `org.hibernate.search.backend.lucene.types.dsl.impl.LuceneStringIndexFieldTypeContext`. I don't see what's different from the Elasticsearch backend?

Note we're just talking about validation here (`hasCompatibleAnalyzer`), not about changing the predicates themselves (`createXXXPredicate`). The only effect would be to trigger a failure if the field happens to use the same Analyzer instance as a normalizer in one index and as an analyzer in the other.

Feel free to merge, but in my opinion this is inconsistent with the ES backend.

fax4ever commented:
Hi @yrodiere, thanks. I wanted to avoid to pass `analyzer` and `normalizer` instances together with the `analyzerOrNormalizer` instance in the constructor of the `LuceneStringIndexFieldTypeContext` as a I've done in the last commit I've just added. Did I get it well?

yrodiere commented:
@fax4ever It looks good; I'll merge once CI goes green.

yrodiere commented:
Merged, thanks!

fax4ever commented:
Thanks!
